src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs

```diff
@@ -22,81 +22,73 @@ namespace Castle.MicroKernel.Lifestyle.Scoped
+	/// Provides explicit lifetime scoping within logical path of execution. Used for types with <see cref="LifestyleType.Scoped" />.
 	/// </summary>
 	/// <remarks>
-	///   The scope is passed on to child threads, including ThreadPool threads. The capability is limited to single
-	///    AppDomain and should be used cautiously as call to <see cref="Dispose" /> may occur while the child thread is still executing, what in turn may lead to subtle threading bugs.
+	/// The scope is passed on to child threads, including ThreadPool threads. The capability is limited to a single AppDomain
+	/// and should be used cautiously as calls to <see cref="Dispose" /> may occur while the child thread is still executing,
+	/// which in turn may lead to subtle threading bugs.
 	/// </remarks>
 	public class CallContextLifetimeScope : ILifetimeScope
 	{
-		private static readonly ConcurrentDictionary<Guid, CallContextLifetimeScope> appDomainLocalInstanceCache =
+		private static readonly ConcurrentDictionary<Guid, CallContextLifetimeScope> allScopes =
 			new ConcurrentDictionary<Guid, CallContextLifetimeScope>();
 
 #if FEATURE_REMOTING
-		private static readonly string keyInCallContext = "castle.lifetime-scope-" + AppDomain.CurrentDomain.Id.ToString(CultureInfo.InvariantCulture);
+		private static readonly string callContextKey = "castle.lifetime-scope-" + AppDomain.CurrentDomain.Id.ToString(CultureInfo.InvariantCulture);
 #else
 		private static readonly AsyncLocal<Guid> asyncLocal = new AsyncLocal<Guid>();
 #endif
-		private readonly Guid contextId;
-		private readonly Lock @lock = Lock.Create();
+
+		private readonly Guid contextId = Guid.NewGuid();
 		private readonly CallContextLifetimeScope parentScope;
+		private readonly Lock @lock = Lock.Create();
 		private ScopeCache cache = new ScopeCache();
 
-		public CallContextLifetimeScope(IKernel container) : this()
-		{
-		}
-
 		public CallContextLifetimeScope()
 		{
-			var parent = ObtainCurrentScope();
-			if (parent != null)
-			{
-				parentScope = parent;
-			}
 			contextId = Guid.NewGuid();
-			var added = appDomainLocalInstanceCache.TryAdd(contextId, this);
+			parentScope = ObtainCurrentScope();
+
+			var added = allScopes.TryAdd(contextId, this);
 			Debug.Assert(added);
 			SetCurrentScope(this);
 		}
 
-		public CallContextLifetimeScope(IWindsorContainer container) : this()
-		{
-		}
-
 		[SecuritySafeCritical]
 		public void Dispose()
 		{
 			using (var token = @lock.ForReadingUpgradeable())
 			{
-				if (cache == null)
-				{
-					return;
-				}
+				// Dispose the burden cache
+				if (cache == null) return;
 				token.Upgrade();
 				cache.Dispose();
 				cache = null;
 
+				// Restore the parent scope (if inside one)
 				if (parentScope != null)
 				{
 					SetCurrentScope(parentScope);
 				}
-#if FEATURE_REMOTING
 				else
 				{
-					CallContext.FreeNamedDataSlot(keyInCallContext);
-				}
+#if FEATURE_REMOTING
```

Comment:
Interesting, why would you let the else potion of this evaluate?

Reply:
It isn't, I just moved the `#ifdef` inside the `else` statement so it just wraps the body.

I was thinking about clearing the `AsyncLocal`'s value then realised that is pointless, and wondered why we were clearing the value from the `CallContext`. Apparently logical call contexts in .NET 4.5+ are now copy-on-write, so we shouldn't need to call `FreeNamedDataSlot` anymore because the entry should disappear as you exit the method that added it. I deferred making the change since it isn't documented on MSDN.